week4/Week4Playground.playground/Contents.swift

```diff
@@ -0,0 +1,154 @@
+import Cocoa
+
+// part (a)
+
+// original array from week 3 part (a)
+// renamed as to not conflict with part (c)
+var arrayOfNums: [Int] = Array(0...20)
+
+// original method from week 3 part (d)
+func average(nums: [Int]?) -> Void {
+    if let nums {
+        let average = nums.reduce(0, +) / nums.count
+        print("The average of the values in the array is \(average).")
+    } else {
+        print("The array is nil. Calculating the average is impossible.")
+    }
+}
+
+func average(nums: [Double]?) -> Double {
+    if let nums {
+        let average = Double(nums.reduce(0, +)) / Double(nums.count)
+        return average
+    } else {
+        return 0.0
+    }
+}
+
+func average(nums: [Int]?, numberOfElements: Int) -> Double {
+    if let nums {
+        let average = Double(nums.reduce(0, +)) / Double(numberOfElements)
+        return average
+    } else {
+        return 0.0
+    }
+}
+
+average(nums: arrayOfNums)
+
+let doubleNums = arrayOfNums.map { Double($0) }
+let avg2 = average(nums: doubleNums)
+let avg3 = average(nums: arrayOfNums, numberOfElements: arrayOfNums.count)
+
+
+// part (b)
+
+enum Animal {
+    case dog, cat, mouse, horse, goat
+}
+
+func theSoundMadeBy(_ animal: Animal) -> String {
+    switch animal {
+    case .dog:
+        return "The dog says woof!"
+    case .cat:
+        return "The cat says meow!"
+    case .mouse:
+        return "The mouse says squeek!"
+    case .horse:
+        return "The horse says ney!"
+    case .goat:
+        return "The goat says ehehehe!"
+    }
+}
+
+theSoundMadeBy(.goat)
+theSoundMadeBy(.mouse)
+
+// part (c)
+
+let nums = Array(0...100)
+let numsWithNil = [79, nil, 80, nil, 90, 100, 72]
+let numsBy2 = Array(2...100).filter { $0 % 2 == 0}
+let numsBy4 = Array(2...100).filter { $0 % 4 == 0}
+
+func evenNumbersArray(from nums: [Int]) -> [Int] {
+    return nums.filter { $0 % 2 == 0 }
+}
+
+print(evenNumbersArray(from: nums))
+
+func sumOfArray(from nums: [Int?]) -> Int {
+    return nums.compactMap { $0 }.reduce(0, +)
+}
+
+print(sumOfArray(from: numsWithNil))
+
+func commonElementsSet(from numsA: [Int], and numsB: [Int]) -> Set<Int> {
+    return Set(numsA.filter { numsB.contains($0) })
```

Comment:
This works but would not be the most efficient method. Think about using the intersection Set method.